secrets/file.go

```diff
@@ -0,0 +1,179 @@
+package secrets
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/fsnotify.v1"
+)
+
+var (
+	ErrAlreadyExists    = errors.New("Secret already exists")
+	ErrWrongFileType    = errors.New("File type not supported")
+	ErrFailedToReadFile = errors.New("Failed to read file")
+)
+
+// SecretsReader is able to get a secret
+type SecretsReader interface {
+	// GetSecret finds secret by name and returns secret and if found or not
+	GetSecret(string) ([]byte, bool)
+}
+
+// SecretsProvider is a SecretsReader and can add secret sources that
+// contain a secret. It will automatically update secrets if the source
+// changed.
+type SecretsProvider interface {
+	SecretsReader
+	// Add adds the given source that contains a secret to the
+	// automatically updated secrets store
+	Add(string) error
+}
+
+type SecretPaths struct {
+	watcher *fsnotify.Watcher
+	known   map[string][]byte
+}
+
+func NewSecretPaths() *SecretPaths {
+	sp := &SecretPaths{
+		known: make(map[string][]byte),
+	}
+
+	if watcher, err := fsnotify.NewWatcher(); err != nil {
+		log.Errorf("Failed to create watcher, probably no support for fsnotify: %v", err)
+	} else {
+		sp.watcher = watcher
+		go sp.startRefresher()
+	}
+
+	return sp
+}
+
+// GetSecret returns secret and if found or not for a given name.
+func (sp *SecretPaths) GetSecret(s string) ([]byte, bool) {
+	dat, ok := sp.known[s]
```

Comment:
the scopes where we access the known field should be protected, because multiple goroutines will use it. An RWMutex is fine.

Reply:
fixed